### PR TITLE
Various Invariant Plotting Fixes

### LINKED ISSUE #1541: Invariant and the infinite multiplication of plots
When using the invariant perspective in 5.0.2, every time the invariant is recomputed a new plot is created (and all the old plots are there as well).  Not sure if the old plots are still linked to the invariant panel or become orphaned.  Thus changing the q limits, the contrast, the background, the extrapolation, the porod constant etc etc etc., each generates a new plot so that quickly I can have a dozen plots floating about.  this should behave as the other perspectives and update the appropriate plot not create a new one (as do inversion and corfunc etc).

Problem noted on both mac and pc

### LINKED ISSUE #1542: Crosstalk between Corfunc and Invariant perspectives
As of 5.0.2, if one sends data to corfunc and plots it, then sends the same data to the invariant all behaves as expected.  However, if in fact the data is sent to Corfunc and the correlation function plotted (hitting `extrapolate` and then `transform`) so that the second plot (the `Real space data`) before then sending the same data set to the invariant, there are two things wrong:
1. the axis labels have now changed from I(Q) vs Q to gamma vs x(A)
2. The plot not only contains the I(Q) data but also 3 lines whose labels suggest they are the `1D Correlation` the `3D Correlation` and the `Interface Distribution Function` which now, as described in issue #1541 get replicated in every plot every time one updates and recomputed the invariant.

NOTE: This can be fixed by sending the data to some other pesrpective and then sending back to the Invariant.

Seen on both Windows  and Mac

### LINKED ISSUE #1605: Problem loading canSAS data into Invariant
The `latex_smeared.xml` data set in the test folder includes 2 separate curves: a slit smeared low Q USANS data set and a high Q pinhole smeared SANS data set.  These are loaded as two separate files into the data manager. Checking only one of the data sets allows only that data to be sent to be plotted or to fitting and once fitted and plotted with the fit only that data set is shown.  however, when sent to the invariant both the checked AND unchecked data sets are plotted in the invariant plot though it appears the Invariant is only calculated from the chosen data set? At least the default qmin and qmax come from that data set and the extrapolations appear to be made from that data set.
NOTE:  This is true in 5.0.2 and in the local build of the 5.x invariant branch.  The dataloader work may or may not have changed this behaviour

### PR DESCRIPTION
**Pull request by krzywon**
This PR fixes a few issues related to plotting, most of which have to do with the Invariant Perspective.
- Fixes #1541: The invariant perspective now updates existing plots so a new window is not generated every time the calculator is run.
- Fixes #1542 and Fixes #1605: The invariant perspective no longer pulls the plot state from the DataExplorer so plots with similar names as Invariant plots are no longer plotted alongside the Invariant plots (same data name or same theory name like Corfunc).

### CODE REVIEW

**Comment by krzywon** [comment]
5.0 ready for testing on Win

File: src/sas/qtgui/MainWindow/DataExplorer.py [hunk 1]
```diff
@@ -1035,7 +1035,7 @@ def displayData(self, data_list, id=None):
         Forces display of charts for the given data set
         """
         # data_list = [QStandardItem, Data1D/Data2D]
-        plot_to_show = data_list[1]
+        plots_to_show = data_list[1:]
```
Comment (rozyczko) [inline src/sas/qtgui/MainWindow/DataExplorer.py:1038]: This is a very good fix not just for Invariant but for plotting in general.

**Review by rozyczko — COMMENTED** [review]
Code looks good.

**Comment by butlerpd** [comment]
- #1541
  - indeed no plots are generated at all if there is no extrapolation.  4.x and before do display the data with no extrapolation as would be expected.
  - If the FIRST extrapolation attempt is only one of the two possiblities, the low Q and the High Q extrapolation seem to generate separate plots, each showing the data and the respective extrapolation.
     -  e.g. If a low q extrapolation alone is done a plot is generated
     -  if then the high Q extrapolation is attempted, either leaving the low q extrapolation on or turning it off generates a second plot showing the low Q extrapolation only.
     - If the high q is left on and the low q extrapolation either turned back on or turned off (depending on which version of the previous step was done) the high Q extrapoltion plot goes blank and does not seem to be recoverable. 
     - The first plot with the low q extrapolation continues to work (seen by making sure the low q extrapolation is on and changing the #points to extrapolate).
  - IF however the first invariant extrapolation attempt uses BOTH low and high then only one plot is generated and only one seems to ever be generated from there on out.
  - There are a variety of variations not all described
- #1542  seems to be fixed from a quick couple of tests
- #1605 more or less fixed
  - only one data set is loaded on graph as expected
  - if both sets in the one file are chosen an error is thrown
  - BUT.. if one choose set a then chooses an extrapolation (in order to generate a plot) and then deselects a and selects b and sends to the invariant (still with the same extrapolation turned on) the data set is NOT updated.  However the extrapolation indicates it is operating on the correct data set.  If one can force a new plot by selecting the other extrapolation, or kills the plot and forces a recalculation, then the new data set does indeed display.  This is probably linked to the effort to suppress the constant generation of new plots?

**Comment by krzywon** [comment]
> * #1541
>   
>   * If the FIRST extrapolation attempt is only one of the two possiblities, the low Q and the High Q extrapolation seem to generate separate plots, each showing the data and the respective extrapolation.

I had also noticed this but thought 2 plots at most is better than 1 for every fit. I partly fix this in PR #1717 (where extrapolation plots are thrown out when checkboxes are unticked) but it's still not completely solved. I can work on this if you think people will find this as (or more) irritating than the current situation. The work around I found is to close the existing plots and recalculate.

> * #1542  seems to be fixed from a quick couple of tests

Excellent!

> * #1605 more or less fixed
>   
>   * BUT.. if one choose set a then chooses an extrapolation (in order to generate a plot) and then deselects a and selects b and sends to the invariant (still with the same extrapolation turned on) the data set is NOT updated.  However the extrapolation indicates it is operating on the correct data set.  If one can force a new plot by selecting the other extrapolation, or kills the plot and forces a recalculation, then the new data set does indeed display.  This is probably linked to the effort to suppress the constant generation of new plots?

I had to test this to see what you're talking about and this is not good... The extrapolations are recalculated, but the plot window is not updated with the new data set. I'll look into this.

**Comment by krzywon** [comment]
My latest commit takes care of the plotting issue related to switching files in the invariant perspective. Each data set gets its own plot when switching data until that plot is closed.

This does not fix the issue of separate plots when extrapolations are calculated separately. I hope to fix that in #1717

**Comment by krzywon** [comment]
My latest set of commits handle the case where the extrapolations are plotted separately. This PR will need a new review at this point.

**Review by rozyczko — APPROVED** [review]
looks good!

**Comment by wpotrzebowski** [comment]
ready for testing on OSX

**Comment by judithhouston** [comment]
ready for testing on OSX

**Comment by smk78** [comment]
@butlerpd commented:
> * #1541
>   * indeed no plots are generated at all if there is no extrapolation.  4.x and before do display the data with no extrapolation as would be expected.

Is this dealt with in any of the PRs?

**Comment by smk78** [comment]
> My latest set of commits handle the case where the extrapolations are plotted separately. This PR will need a new review at this point.

Building this branch locally I find:
* Only enabling low-Q ext > Calculate, only generates one plot
* Then enabling high-Q ext > Calculate, adds that to the existing plot
* Unchecking either ext > Calculate, does nothing but that is not part of this PR
* Only enabling high-Q ext > Calculate, only generates one plot
* Then enabling low-Q ext > Calculate, adds that to the existing plot
* Unchecking either ext > Calculate, does nothing again

So functionality would seem to be as intended.

**Comment by smk78** [comment]
@butlerpd commented:
> > * #1605 more or less fixed
> >   
> >   * BUT.. if one choose set a then chooses an extrapolation (in order to generate a plot) and then deselects a and selects b and sends to the invariant (still with the same extrapolation turned on) the data set is NOT updated.  However the extrapolation indicates it is operating on the correct data set.  If one can force a new plot by selecting the other extrapolation, or kills the plot and forces a recalculation, then the new data set does indeed display.  This is probably linked to the effort to suppress the constant generation of new plots?

@krzywon replied:
> > I had to test this to see what you're talking about and this is not good... The extrapolations are recalculated, but the plot window is not updated with the new data set. I'll look into this.
>
> My latest commit takes care of the plotting issue related to switching files in the invariant perspective. Each data set gets its own plot when switching data until that plot is closed.

Building this branch locally I find:

* Load set A > Invariant > select an Ext > Calculate, generates one plot as above
* Load Set B > _deselect Set A_ > Invariant > select an Ext > Calculate, generates a second plot and switches the dataset name in the Invariant panel
* Load Set C with plots from Set A and Set B still up > _deselect Set A and Set B_ > Invariant > select an Ext > Calculate, generates a third plot

So functionality would seem to be as intended.

ALTHOUGH:
* Deselect Set B > reselect Set A > Invariant > select an Ext > Calculate, switches the dataset name back in the Invariant panel and does nothing else, or adds the data and the calculated extrapolation to the plot for Set B, not Set A. _However at the moment this behaviour is quite tricky to reproduce because this branch doesn't have enough functionality to help you force a redraw!_

**Review by smk78 — APPROVED** [review]
Have reviewed for functionality and I think is good to merge at this stage.

**Pull request merged**